### DISCUSSION TITLE
Extension sponsorship

### DISCUSSION THREAD

**Discussion by isidorn** [microsoft/vscode-discussions#14] [Announcements]
We have introduced a `sponsor` field in the extension's `package.json` to allow extensions to opt into sponsorship.
If an extension opts-into this VS Code will render a Sponsor button in the extension details page. 
We might add more UI cues in the future on VS Code side. For Marketplace we have this issue to track the work https://github.com/microsoft/vsmarketplace/issues/398

We hope this will allow our users to fund the extensions that they depend on to improve the extension's performance, reliability, and stability. 

---
Example:

An extension can add the following to their `package.json`
```
"sponsor": {
		"url": "https://github.com/sponsors/nvaccess"
}
```

Make sure to use latest vsce `2.9.1` when publishing your extension.

My tryout extension demonstrating this:

![Screenshot 2022-06-30 at 12 36 32](https://user-images.githubusercontent.com/1926584/176657138-aae79cc2-887c-4bd8-8ef2-c4289041c85c.png)

Once users clicks on ❤️  Sponsor button she is taken to the [sponsor link](https://github.com/sponsors/nvaccess) my extension provided.

Try it out and let us know what you think. Thanks!

**Comment by PEZ** [comment]
I've added this to [Calva 2.0.281](https://marketplace.visualstudio.com/items?itemName=betterthantomorrow.calva). Though I don't see the button in VS Code. Nor in insiders. Did I do it wrong?

https://github.com/BetterThanTomorrow/calva/blob/published/package.json#L21-L23

**Comment by isidorn** [comment]
@pez make sure to use latest vsce `2.9.1` when publishing your extension. I will make this more clear in my comment above.

Also the button will right now only be rendered in Insiders. In VS Code stable it should be shown when we release end of next week.

**Reply by PEZ** [reply]
Thanks! I still can't get it to work, however. For a long while it seemed to be because of a bug in npx where it refuses to use any other version than what it has cached somewhere. But even installing vsce 2.9.1 globally and using that, I still don't see a button.

Here's a VSIX I packaged on my machine. 

[calva-2.0.282-sponsors.zip](https://github.com/microsoft/vscode-discussions/files/8835561/calva-2.0.282-sponsors.zip)

Is there something missing in it? Maybe I need to update my `.vscodeignore`?

**Comment by isidorn** [comment]
@pez thanks for doing this. I just tried from my VS Code Insiders and I see the sponsor button nicely rendered. Does it work for you now?

![Screenshot 2022-06-06 at 09 59 00](https://user-images.githubusercontent.com/1926584/172120328-9865f8b2-f56a-48c7-a5b7-ecc1bc3d8114.png)

**Reply by PEZ** [reply]
Checking, and I have the button now. That's great!

**Comment by alefragnani** [comment]
Does it work only with _published_ extensions?

I just tried installing a `.vsix` file (generated with `vsce package` command) and it is not showing the button 😕

I'm using `vsce` version 2.9.1 and VS Code Insiders  build from yesterday (June 7th)

**Comment by isidorn** [comment]
@alefragnani I just installed a Calva vsix from https://marketplace.visualstudio.com/items?itemName=betterthantomorrow.calva and the sponsor button is nicely shown.
It might be that `vsce publish` has to be called, in order for the sponsor tag to be added which we use to decide if we should show the sponsor button or not.

**Reply by alefragnani** [reply]
Hm, I see.

I thought the `package` command would be the responsible for that, not the `publish`. For me, the `publish` was responsible for uploading to the Marketplace only. Good to know...

BTW, do you know how VS Code _alternatives_ like Gitpod and VSCodium would work? Since these projects can't rely on the VS Marketplace backend, I now wonder if this would work.

Thank you

**Reply by isidorn** [reply]
I just checked with engineering and we actually do it as part of the `package`. So my previous statement was wrong.
I am not sure why it was not working for you, sorry.

For VSCodium and Gitpod I am not sure how this would work, sorry.

**Reply by alefragnani** [reply]
Just to give you a feedback, I _published_ a temporary release and it worked just fine. So for whatever reason, it only works on _published_ (maybe some dependency/issue with the VS Marketplace service)

Rolled back for now until my sponsors profile is up and running.

**Comment by formulahendry** [comment]
https://marketplace.visualstudio.com/items?itemName=formulahendry.code-runner has adopted it. Thanks a lot!! ❤️ 

![image](https://user-images.githubusercontent.com/1050213/173195287-a880955a-c7fb-4cb0-a531-ca2725d8b210.png)

**Comment by nickserv** [comment]
How about using the `funding` property for consistency with npm?

**Reply by isidorn** [reply]
Good idea but I was not aware of the name in `npm`, and changing the name now would break existing extensions that use it.
So let's still use "sponsor". But thank you!

**Reply by bwateratmsft** [reply]
Couldn't the extension view be configured to understand both (to avoid breaking existing extensions), but only document/recommend the `funding` tag?

**Reply by isidorn** [reply]
Good idea and yes it would be possible. Just not sure if it is worth the effort.

**Comment by filiptronicek** [comment]
Hi @isidorn 👋! 
Have you explored adding multiple links for sponsorships? It would be nice to have feature parity / similar behavior to GitHub's [FUNDING.yml](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/displaying-a-sponsor-button-in-your-repository).

**Comment by filiptronicek** [comment]
I also had an idea in that there could be a bit more interactivity with GitHub Sponsors - the Sponsor button could become "Sponsoring" if the logged-in GitHub user sponsors that user / organization with the help of the [GitHub Sponsors GraphQL API](https://github.com/github-community/community/discussions/3818).

It's most likely overkill just for one nice button, but it would be a nice touch 😊.

**Reply by nickserv** [reply]
Adding to this, if the extension's `repository` field uses GitHub, it would be great to use its sponsorship configuration automatically.

**Reply by isidorn** [reply]
Great ideas. Feel free to open a new feature request in VS Code for this.
However I think it will probably be a bit too much work and not something we would do in the near future.

**Comment by alefragnani** [comment]
Updated all my extensions (https://marketplace.visualstudio.com/publishers/alefragnani) to support it! Thank you for the feature.

![Screen Shot 2022-07-18 at 18 19 05](https://user-images.githubusercontent.com/3781424/179619238-8edeedbf-b23e-490a-a8b5-0a1a2c457e46.png)

Now lets just wait for users to click on that button 😬

**Comment by isidorn** [comment]
Hi Extension authors,

I want to check if you noticed any noticeable change with the number of sponsorships your extensions are receiving since we introduced the Sponsor feature in VS Code?

Thanks!

**Reply by alefragnani** [reply]
I can’t say for sure, but my feeling is yes!. That’s because my Sponsors profile was launched only after the VS Code feature was available. So, in the end, both came to life together.

But, if I compare the numbers of sponsorships I received here, with the number I reached in Patreon, it was a great improvement. In less than two months, I got the same number of twelve months in Patreon. 

Thank you!

**Reply by PEZ** [reply]
My datapoint is probably not representative, so I hesitate to submit it. I'd say I haven't seen much difference. I remain super thankful for this feature and am hoping you will iterate on it.

What I think makes Calva special here is that it has very few users in comparison to most VS Code extension with any similar scope of features. Clojure is fringe. It also has a very engaged community. So I get to communicate with many of my users. Whoever is ready to sponsor the project will probably find reasons to do so in those interactions. I'm sure they might also get reminded by seeing that heart in VS Code itself as well, but I think that for many other extensions that effect is probably much stronger.

I love this feature, and I think it is beautifully implemented. It is also a very nice nod to open source, and the people who are involved in it. As is VS Code itself.
